### PR TITLE
select many2one field value only if floating

### PR DESCRIPTION
**Pull request by msh-odoo**
PURPOSE
When tabbing out of a focus m2x field, the 'most likely record' is set > this is a good thing
problem is, we only want to do that if the user actually wrote something

SPEC
Only set the first value of the dropdown is the user wrote at least one character

TASK 2455781

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/tests/fields/relational_fields/field_many2one_tests.js [hunk 1]
```diff
@@ -2721,6 +2721,41 @@ QUnit.module('fields', {}, function () {
...
+            const $input = form.$('.o_field_many2one input');
+            await testUtils.dom.click($input);
+            await testUtils.fields.triggerKeydown($input, 'tab');
+            assert.strictEqual($input.val(), '', "no record should have been selected");
+
+            // open autocomplete dropdown and manually select item by UP/DOWN key and press TAB
+            await testUtils.dom.click($input);
+            await testUtils.fields.triggerKeydown($input, 'down');
+            await testUtils.fields.triggerKeydown($input, 'tab');
+            assert.strictEqual($input.val(), 'second record', "no record should have been selected");
```
Comment (FrancoisGe) [inline addons/web/static/tests/fields/relational_fields/field_many2one_tests.js:2746]: no record should have been selected ???
Maybe: "second record should have been selected " ?

File: addons/web/static/tests/fields/relational_fields/field_many2one_tests.js [hunk 2]
```diff
@@ -2721,6 +2721,41 @@ QUnit.module('fields', {}, function () {
...
+
+            // clear many2one and then open autocomplete, write something and press TAB
+            await testUtils.fields.editAndTrigger($input, '', 'keydown');
+            await testUtils.dom.click($input);
+            await testUtils.dom.click($input);
+            await testUtils.fields.editAndTrigger($input, 'aa', 'keyup');
+            await testUtils.fields.triggerKeydown($input, 'tab');
+            assert.strictEqual($input.val(), 'first record', "no record should have been selected");
```
Comment (FrancoisGe) [inline addons/web/static/tests/fields/relational_fields/field_many2one_tests.js:2754]: no record should have been selected ???
Maybe: "first record should have been selected " ?

**Review by FrancoisGe — COMMENTED** [review]
Good work. Could you please fix my remarks?

File: addons/web/static/tests/fields/relational_fields/field_many2one_tests.js [hunk 3]
```diff
@@ -2721,6 +2721,41 @@ QUnit.module('fields', {}, function () {
...
+            // open autocomplete dropdown and manually select item by UP/DOWN key and press TAB
+            await testUtils.dom.click($input);
+            await testUtils.fields.triggerKeydown($input, 'down');
+            await testUtils.fields.triggerKeydown($input, 'tab');
+            assert.strictEqual($input.val(), 'second record', "no record should have been selected");
+
+            // clear many2one and then open autocomplete, write something and press TAB
+            await testUtils.fields.editAndTrigger($input, '', 'keydown');
+            await testUtils.dom.click($input);
+            await testUtils.dom.click($input);
```
Comment (brboi) [inline addons/web/static/tests/fields/relational_fields/field_many2one_tests.js:2751]: Why click twice ?

File: addons/web/static/tests/fields/relational_fields/field_many2one_tests.js [hunk 4]
```diff
@@ -2722,6 +2717,67 @@ QUnit.module('fields', {}, function () {
...
+                        <field name="display_name"/>
+                    </form>`,
+                res_id: 1,
+                viewOptions: {
+                    mode: 'edit',
+                },
+            });
+
+            const $input = form.$('.o_field_many2one input');
+            await testUtils.fields.editInput($input, "");
```
Comment (aab-odoo) [inline addons/web/static/tests/fields/relational_fields/field_many2one_tests.js:2772]: could you assert that the input isn't empty at the beginning?

File: addons/web/static/tests/fields/relational_fields/field_many2one_tests.js [hunk 5]
```diff
@@ -2722,6 +2717,67 @@ QUnit.module('fields', {}, function () {
...
+            await testUtils.fields.triggerKeydown($input, 'down');
+            await testUtils.fields.triggerKeydown($input, 'tab');
+            assert.strictEqual($input.val(), 'second record', "second record should have been selected");
+
+            // clear many2one and then open autocomplete, write something and press TAB
+            await testUtils.fields.editAndTrigger(form.$('.o_field_many2one input'), '', ['keyup', 'blur']);
+            await testUtils.dom.click($input);
+            await testUtils.fields.editAndTrigger($input, 'aa', 'keyup');
+            await testUtils.fields.triggerKeydown($input, 'tab');
+            assert.strictEqual($input.val(), 'first record', "first record should have been selected");
```
Comment (aab-odoo) [inline addons/web/static/tests/fields/relational_fields/field_many2one_tests.js:2749]: this is weird, you type 'aa' and it selects 'first record'?

**Review by aab-odoo — COMMENTED** [review]
I made a few remarks on the test. Also, please check the video I posted on the task. Thanks

**Comment by aab-odoo** [comment]
robodoo r+

**Comment by aab-odoo** [comment]
robodoo retry

**Comment by msh-odoo** [comment]
@aab-odoo Branch was red due to which merging failed, would you please try now?

**Comment by aab-odoo** [comment]
robodoo r+

**Pull request closed without merging**